test/rules/invalid-void/test.ts.lint

```diff
@@ -0,0 +1,34 @@
+function func(): void {}
+
+type NormalType = () => void;
+
+interface Interface {
+    lambda: () => void;
+    voidProp: void;
+              ~~~~ [void as non-return type is forbidden]
+}
+
+class ClassName {
+    private readonly propName: void;
+                               ~~~~ [void as non-return type is forbidden]
+}
+
+let letVoid: void;
+             ~~~~ [void as non-return type is forbidden]
+
+type VoidType = void;
+                ~~~~ [void as non-return type is forbidden]
+
+class OtherClassName {
+    private propName: VoidType;
+}
+
+type UnionType = string | number;
+type UnionType2 = string | number | void;
+                                    ~~~~ [void as non-return type is forbidden]
+
+type UnionType3 = string | (number & any | (string | void));
+                                                     ~~~~ [void as non-return type is forbidden]
+
+type IntersectionType = string & number & void;
+                                          ~~~~ [void as non-return type is forbidden]
```

Comment:
What do you mean? Change the default value in `Lint.WalkContext` from `void` to `{}`?

Reply:
I'm not sure what did you mean, but I've added new kinds to cover cases you pointed out in other comment and now `Lint.WalkContext` is marked as "invalid" and I changed default for generic type from `void` to `{}` for both `WalkContext` and `AbstractWalker` in c8ff62ed9c31c945f387d68d4f44ef5a265472f0.

Reply:
Oh hey sorry - missed this comment.

`void` is a weird and spooky value for a generic. Most uses of generics are to represent the stored value in some place, such as:

```ts
type Box<T> = {
    value: T;
};

let numberBox: Box<number> = { value: 7, };
```

`Lint.WalkContext` is a more complex form of that. Options are stored as the generic type, which is the equivalent of `value: T;` from that example. Using generics to say the stored type is `void` is weird because, per this rule, the `void` type shouldn't be used for variable storage.

One could argue for an option in to allow generics to be `void`, in the case of them being used as return values:

```ts
type BoxCreator<T> = () => T;
```

...but that seems like it should be discussed in a separate issue so as to not block this PR. Why would we want a return value that's only sometimes useful? Spooky. 👻

Reply:
Linking this to https://github.com/palantir/tslint/pull/4736#discussion_r300407267: although `void` isn't a valid storage type, `undefined` is.

Reply:
> One could argue for an option in to allow generics to be void, in the case of them being used as return values:

Interesting how many cases is there of that 🙂 They are, but how many in comparison with other generics (specially with possible invalid usage of `void` there). Anyway, I believe that cases can be easily "fixed" with disabling this rule locally.